Conditional PackageReference version splitting analyzer presence fails

If a project references two different versions of the same package in two different TargetFrameworks, and the package has added or removed a Roslyn analyzer, the analyzer will be added from _both_ versions of the package, even the one where it's not present.

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>net6.0;net472</TargetFrameworks>
    <RootNamespace>two_systemtextjson</RootNamespace>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Text.Json" Version="4.7.0" Condition=" '$(TargetFramework)' == 'net472' " />
    <PackageReference Include="System.Text.Json" Version="6.0.0-preview.4.21253.7" Condition=" '$(TargetFramework)' == 'net6.0' " />
  </ItemGroup>

</Project>
```

```sh-session
two-systemtextjson on master .NET v6.0.100-preview.5.21302.13 🎯 net6.0;net472 
❯ dotnet build  
Microsoft (R) Build Engine version 17.0.0-preview-21302-02+018bed83d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored S:\work\two-systemtextjson\two-systemtextjson.csproj (in 135 ms).
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
CSC : error CS0006: Metadata file 'C:\Users\raines\.nuget\packages\system.text.json\4.7.0\analyzers\dotnet\cs\System.Text.Json.SourceGeneration.dll' could not be found [S:\work\two-systemtextjson\two-systemtextjson.csproj]
  two-systemtextjson -> S:\work\two-systemtextjson\bin\Debug\net6.0\two-systemtextjson.dll

Build FAILED.

CSC : error CS0006: Metadata file 'C:\Users\raines\.nuget\packages\system.text.json\4.7.0\analyzers\dotnet\cs\System.Text.Json.SourceGeneration.dll' could not be found [S:\work\two-systemtextjson\two-systemtextjson.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.20
```

Because 4.7.0 of that package didn't have an analyzer/source generator, the file isn't found in that package version. The reference is only there because it was found in the other version.

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
@@ -14,6 +14,7 @@ using Microsoft.Build.Utilities;
 using NuGet.Common;
 using NuGet.Frameworks;
 using NuGet.ProjectModel;
+using NuGet.Versioning;
 
 namespace Microsoft.NET.Build.Tasks
 {
@@ -824,7 +825,7 @@ namespace Microsoft.NET.Build.Tasks
 
             private void WriteAnalyzers()
             {
-                Dictionary<string, LockFileTargetLibrary> targetLibraries = null;
+                Dictionary<(string, NuGetVersion), LockFileTargetLibrary> targetLibraries = null;
 
                 foreach (var library in _lockFile.Libraries)
                 {
@@ -842,12 +843,10 @@ namespace Microsoft.NET.Build.Tasks
 
                         if (targetLibraries == null)
                         {
-                            targetLibraries = _runtimeTarget
-                                .Libraries
-                                .ToDictionary(l => l.Name, StringComparer.OrdinalIgnoreCase);
+                            targetLibraries = _compileTimeTarget.Libraries.ToDictionary(l => (l.Name, l.Version));
                         }
 
-                        if (targetLibraries.TryGetValue(library.Name, out var targetLibrary))
+                        if (targetLibraries.TryGetValue((library.Name, library.Version), out var targetLibrary))
                         {
                             WriteItem(_packageResolver.ResolvePackageAssetPath(targetLibrary, file), targetLibrary);
                         }

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
@@ -833,9 +833,16 @@ namespace Microsoft.NET.Build.Tasks
                 }
                 public int GetHashCode((string, NuGetVersion) library)
                 {
+#if NET
                     return HashCode.Combine(
                         StringComparer.OrdinalIgnoreCase.GetHashCode(library.Item1),
                         library.Item2.GetHashCode());
+#else
+                    int hashCode = -1507694697;
+                    hashCode = hashCode * -1521134295 + StringComparer.OrdinalIgnoreCase.GetHashCode(library.Item1);
+                    hashCode = hashCode * -1521134295 + library.Item2.GetHashCode();
+                    return hashCode;
+#endif
                 }
             }
 

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
@@ -823,6 +823,22 @@ namespace Microsoft.NET.Build.Tasks
                 Position = savedPosition;
             }
 
+            class LibraryComparer : IEqualityComparer<(string, NuGetVersion)>
+            {
+                public bool Equals((string, NuGetVersion) l1, (string, NuGetVersion) l2)
+                {
+                    return StringComparer.OrdinalIgnoreCase.Equals(l1.Item1, l2.Item1)
+                        && l1.Item2.Equals(l2.Item2);
+                    
+                }
+                public int GetHashCode((string, NuGetVersion) library)
+                {
+                    return HashCode.Combine(
+                        StringComparer.OrdinalIgnoreCase.GetHashCode(library.Item1),
+                        library.Item2.GetHashCode());
+                }
+            }
+
             private void WriteAnalyzers()
             {
                 Dictionary<(string, NuGetVersion), LockFileTargetLibrary> targetLibraries = null;
@@ -843,7 +859,7 @@ namespace Microsoft.NET.Build.Tasks
 
                         if (targetLibraries == null)
                         {
-                            targetLibraries = _compileTimeTarget.Libraries.ToDictionary(l => (l.Name, l.Version));
+                            targetLibraries = _compileTimeTarget.Libraries.ToDictionary(l => (l.Name, l.Version), new LibraryComparer());
                         }
 
                         if (targetLibraries.TryGetValue((library.Name, library.Version), out var targetLibrary))

--- a/src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToUseAnalyzers.cs
+++ b/src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToUseAnalyzers.cs
@@ -13,6 +13,7 @@ using System.Collections.Generic;
 using System;
 using Xunit.Abstractions;
 using System.Runtime.InteropServices;
+using Microsoft.NET.TestFramework.ProjectConstruction;
 
 namespace Microsoft.NET.Build.Tests
 {
@@ -55,25 +56,6 @@ namespace Microsoft.NET.Build.Tests
 
             var analyzers = command.GetValues();
 
-            List<string> nugetRoots = new List<string>()
-            {
-                TestContext.Current.NuGetCachePath,
-                Path.Combine(FileConstants.UserProfileFolder, ".dotnet", "NuGetFallbackFolder")
-            };
-
-            string RelativeNuGetPath(string absoluteNuGetPath)
-            {
-                foreach (var nugetRoot in nugetRoots)
-                {
-                    if (absoluteNuGetPath.StartsWith(nugetRoot + Path.DirectorySeparatorChar))
-                    {
-                        return absoluteNuGetPath.Substring(nugetRoot.Length + 1)
-                                    .Replace(Path.DirectorySeparatorChar, '/');
-                    }
-                }
-                throw new InvalidDataException("Expected path to be under a NuGet root: " + absoluteNuGetPath);
-            }
-
             switch (language)
             {
                 case "C#":
@@ -103,6 +85,70 @@ namespace Microsoft.NET.Build.Tests
                 default:
                     throw new ArgumentOutOfRangeException(nameof(language));
             }
+        }
+
+        [Fact]
+        public void It_resolves_multitargeted_analyzers()
+        {
+            var testProject = new TestProject()
+            {
+                TargetFrameworks = "net6.0;net472"
+            };
+
+            //  Disable analyzers built in to the SDK so we can more easily test the ones coming from NuGet packages
+            testProject.AdditionalProperties["EnableNETAnalyzers"] = "false";
+            testProject.AdditionalProperties["DisableFrameworkReferenceAnalyzers"] = "false";
+
+            testProject.ProjectChanges.Add(project =>
+            {
+                var ns = project.Root.Name.Namespace;
+
+                var itemGroup = XElement.Parse(@"
+  <ItemGroup>
+    <PackageReference Include=""System.Text.Json"" Version=""4.7.0"" Condition="" '$(TargetFramework)' == 'net472' "" />
+    <PackageReference Include=""System.Text.Json"" Version=""6.0.0-preview.4.21253.7"" Condition="" '$(TargetFramework)' == 'net6.0' "" />
+  </ItemGroup>");
+
+                project.Root.Add(itemGroup);
+            });
+
+            var testAsset = _testAssetsManager.CreateTestProject(testProject);
+
+            List<string> GetAnalyzersForTargetFramework(string targetFramework)
+            {
+                var getValuesCommand = new GetValuesCommand(testAsset,
+                    valueName: "Analyzer",
+                    GetValuesCommand.ValueType.Item,
+                    targetFramework);
+
+                getValuesCommand.DependsOnTargets = "ResolveLockFileAnalyzers";
+
+                getValuesCommand.Execute("-p:TargetFramework=" + targetFramework).Should().Pass();
+
+                return getValuesCommand.GetValues().Select(RelativeNuGetPath).ToList();
+            }
+            
+            GetAnalyzersForTargetFramework("net6.0").Should().BeEquivalentTo("system.text.json/6.0.0-preview.4.21253.7/analyzers/dotnet/cs/System.Text.Json.SourceGeneration.dll");
+            GetAnalyzersForTargetFramework("net472").Should().BeEmpty();
+        }
+
+        static readonly List<string> nugetRoots = new List<string>()
+            {
+                TestContext.Current.NuGetCachePath,
+                Path.Combine(FileConstants.UserProfileFolder, ".dotnet", "NuGetFallbackFolder")
+            };
+
+        static string RelativeNuGetPath(string absoluteNuGetPath)
+        {
+            foreach (var nugetRoot in nugetRoots)
+            {
+                if (absoluteNuGetPath.StartsWith(nugetRoot + Path.DirectorySeparatorChar))
+                {
+                    return absoluteNuGetPath.Substring(nugetRoot.Length + 1)
+                                .Replace(Path.DirectorySeparatorChar, '/');
+                }
+            }
+            throw new InvalidDataException("Expected path to be under a NuGet root: " + absoluteNuGetPath);
         }
     }
 }

